tests/test_single_topology_v3.py

```diff
@@ -0,0 +1,217 @@
+    all_idxs, _ = setup_dummy_interactions_from_ff(ff, mol, dummy_group=[6, 12], root_anchor_atom=5, nbr_anchor_atom=0)
+    bond_idxs, angle_idxs, improper_idxs = all_idxs
+
+    assert set(bond_idxs) == set([(5, 6), (6, 12)])
+    assert set(angle_idxs) == set([(5, 6, 12), (0, 5, 6)])
+    assert set(improper_idxs) == set()
+
+    # set [H] as the dummy group, without neighbor anchor atom
+    all_idxs, _ = setup_dummy_interactions_from_ff(ff, mol, dummy_group=[12], root_anchor_atom=6, nbr_anchor_atom=None)
+    bond_idxs, angle_idxs, improper_idxs = all_idxs
+
+    assert set(bond_idxs) == set([(6, 12)])
+    assert set(angle_idxs) == set()
+    assert set(improper_idxs) == set()
+
+    # set [H] as the dummy group, with neighbor anchor atom
+    all_idxs, _ = setup_dummy_interactions_from_ff(ff, mol, dummy_group=[12], root_anchor_atom=6, nbr_anchor_atom=5)
+    bond_idxs, angle_idxs, improper_idxs = all_idxs
+
+    assert set(bond_idxs) == set([(6, 12)])
+    assert set(angle_idxs) == set([(5, 6, 12)])
+    assert set(improper_idxs) == set()
+
+    with pytest.raises(single_topology_v3.MissingAngleError):
+        all_idxs, _ = setup_dummy_interactions_from_ff(ff, mol, dummy_group=[12], root_anchor_atom=6, nbr_anchor_atom=4)
+
+
+def test_find_dummy_groups_and_anchors():
+    """
+    Test that we can find the anchors and dummy groups when there's a single core anchor atom. When core bond
+    is broken, we should disable one of the angle atoms.
+    """
+    mol_a = Chem.MolFromSmiles("OCCC")
+    mol_b = Chem.MolFromSmiles("CCCF")
+    core_pairs = np.array([[1, 2], [2, 1], [3, 0]])
+
+    dgs, jks = single_topology_v3.find_dummy_groups_and_anchors(mol_a, mol_b, core_pairs[:, 0], core_pairs[:, 1])
+    assert dgs == [{3}]
+    assert jks == [(2, 1)]
+
+    # angle should swap
+    core_pairs = np.array([[1, 2], [2, 0], [3, 1]])
+
+    with pytest.warns(CoreBondChangeWarning):
+        dgs, jks = single_topology_v3.find_dummy_groups_and_anchors(mol_a, mol_b, core_pairs[:, 0], core_pairs[:, 1])
+        assert dgs == [{3}]
+        assert jks == [(2, None)]
+
+
+def test_find_dummy_groups_and_anchors_multiple_angles():
+    """
+    Test that when multiple angle groups are possible we can find one deterministically
+    """
+    mol_a = Chem.MolFromSmiles("CCC")
+    mol_b = Chem.MolFromSmiles("CC(C)C")
+
+    core_pairs = np.array([[0, 2], [1, 1], [2, 3]])
+    dgs, jks = single_topology_v3.find_dummy_groups_and_anchors(mol_a, mol_b, core_pairs[:, 0], core_pairs[:, 1])
+    assert dgs == [{0}]
+    assert jks == [(1, 2)] or jks == [(1, 3)]
+
+    dgs_zero, jks_zero = single_topology_v3.find_dummy_groups_and_anchors(
+        mol_a, mol_b, core_pairs[:, 0], core_pairs[:, 1]
+    )
+
+    # this code should be invariant to different random seeds and different ordering of core pairs
+    for idx in range(100):
+        np.random.seed(idx)
+        core_pairs_shuffle = np.array(core_pairs)
```

Comment:
(Nit: don't forget to shuffle these if intended! Currently may just create a copy)

```suggestion
        np.random.seed(idx)
        core_pairs_shuffle = np.array(core_pairs)
        np.random.shuffle(core_pairs_shuffle)
```